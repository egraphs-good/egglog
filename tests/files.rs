use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
