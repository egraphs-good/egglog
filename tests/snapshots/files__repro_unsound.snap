---
source: tests/files.rs
expression: serialized
---
nodes:
  Flip__-16212409053862882115:
    op: Flip__
    children: []
    eclass: "121"
    cost: 1
  Flip__-10444575304181264970:
    op: Flip__
    children: []
    eclass: "164"
    cost: 1
  Flip__-13926100405575019960:
    op: Flip__
    children: []
    eclass: "218"
    cost: 1
  Flip__-13146672979998668357:
    op: Flip__
    children: []
    eclass: "243"
    cost: 1
  Flip__-8210240284334595469:
    op: Flip__
    children: []
    eclass: "283"
    cost: 1
  Flip__-779604426617889810:
    op: Flip__
    children: []
    eclass: "332"
    cost: 1
  Flip__-6703359061623084917:
    op: Flip__
    children: []
    eclass: "355"
    cost: 1
  Flip__-11795712642610735767:
    op: Flip__
    children: []
    eclass: "381"
    cost: 1
  Flip__-7067174460711433495:
    op: Flip__
    children: []
    eclass: "509"
    cost: 1
  Flip__-8002522771611363060:
    op: Flip__
    children: []
    eclass: "550"
    cost: 1
  TrmOf__-9301332479516564789:
    op: TrmOf__
    children:
      - Repi64__-12574963067145549163
    eclass: "31"
    cost: 1000000
  TrmOf__-10964134587551653303:
    op: TrmOf__
    children:
      - Repi64__-5040379952546458196
    eclass: "25"
    cost: 1000000
  TrmOf__-3481525101393754990:
    op: TrmOf__
    children: []
    eclass: "52"
    cost: 1000000
  TrmOf__-12626936695586741817:
    op: TrmOf__
    children:
      - Repi64__-17615343019692007359
    eclass: "19"
    cost: 1000000
  TrmOf__-14289738803621830331:
    op: TrmOf__
    children:
      - Repi64__-11743562013128004906
    eclass: "13"
    cost: 1000000
  TrmOf__-15952540911656918845:
    op: TrmOf__
    children: []
    eclass: "7"
    cost: 1000000
  TrmOf__-8469931425499020532:
    op: TrmOf__
    children: []
    eclass: "34"
    cost: 1000000
  TrmOf__-17615343019692007359:
    op: TrmOf__
    children:
      - Repi64__-0
    eclass: "1"
    cost: 1000000
  TrmOf__-13562285006486671382:
    op: TrmOf__
    children: []
    eclass: "60"
    cost: 1000000
  TrmOf__-10132733533534109046:
    op: TrmOf__
    children: []
    eclass: "28"
    cost: 1000000
  i64-0:
    op: "0"
    children: []
    eclass: i64-0
    cost: 0
  Repi64__-0:
    op: Repi64__
    children:
      - i64-0
    eclass: "3"
    cost: 1
  i64-5871781006564002453:
    op: "1"
    children: []
    eclass: i64-5871781006564002453
    cost: 0
  Repi64__-5871781006564002453:
    op: Repi64__
    children:
      - i64-5871781006564002453
    eclass: "6"
    cost: 1
  i64-11743562013128004906:
    op: "2"
    children: []
    eclass: i64-11743562013128004906
    cost: 0
  Repi64__-11743562013128004906:
    op: Repi64__
    children:
      - i64-11743562013128004906
    eclass: "15"
    cost: 1
  i64-17615343019692007359:
    op: "3"
    children: []
    eclass: i64-17615343019692007359
    cost: 0
  Repi64__-17615343019692007359:
    op: Repi64__
    children:
      - i64-17615343019692007359
    eclass: "21"
    cost: 1
  i64-5040379952546458196:
    op: "4"
    children: []
    eclass: i64-5040379952546458196
    cost: 0
  Repi64__-5040379952546458196:
    op: Repi64__
    children:
      - i64-5040379952546458196
    eclass: "27"
    cost: 1
  i64-12574963067145549163:
    op: "-1"
    children: []
    eclass: i64-12574963067145549163
    cost: 0
  Repi64__-12574963067145549163:
    op: Repi64__
    children:
      - i64-12574963067145549163
    eclass: "33"
    cost: 1
  Rational-11743562013128004906:
    op: (rational 2 1)
    children: []
    eclass: Rational-11743562013128004906
    cost: 0
  r-two-0:
    op: r-two
    children: []
    eclass: Rational-11743562013128004906
    cost: 9223372036854776000
  AstIf_HerbieType_Math_Math_Math__-12757547884481145869:
    op: AstIf_HerbieType_Math_Math_Math__
    children:
      - GetChild__-17074017086827668002
      - TrmOf__-3481525101393754990
    eclass: "123"
    cost: 1
  GetChild__-1465098945834158911:
    op: GetChild__
    children:
      - TrmOf__-13562285006486671382
      - i64-0
    eclass: "41"
    cost: 1000000
  GetChild__-14040062012979708074:
    op: GetChild__
    children:
      - AstExpm1_HerbieType_Math__-15564543775964812645
      - i64-5871781006564002453
    eclass: "48"
    cost: 1000000
  GetChild__-17074017086827668002:
    op: GetChild__
    children:
      - AstExpm1_HerbieType_Math__-2158179654801719225
      - i64-0
    eclass: "41"
    cost: 1000000
  GetChild__-4499054019682118839:
    op: GetChild__
    children:
      - AstExpm1_HerbieType_Math__-2158179654801719225
      - i64-5871781006564002453
    eclass: "52"
    cost: 1000000
  GetChild__-17279319512441093813:
    op: GetChild__
    children:
      - GetChild__-1465098945834158911
      - i64-0
    eclass: "37"
    cost: 1000000
  GetChild__-5706686836758916083:
    op: GetChild__
    children:
      - TrmOf__-15952540911656918845
      - i64-0
    eclass: "1"
    cost: 1000000
  GetChild__-18281649903904465246:
    op: GetChild__
    children:
      - TrmOf__-15952540911656918845
      - i64-5871781006564002453
    eclass: "4"
    cost: 1000000
  GetChild__-5145996267849227305:
    op: GetChild__
    children:
      - i64-0
    eclass: "4"
    cost: 1000000
  GetChild__-17720959334994776468:
    op: GetChild__
    children:
      - i64-5871781006564002453
    eclass: "4"
    cost: 1000000
  GetChild__-15002184148634740066:
    op: GetChild__
    children:
      - TrmOf__-8469931425499020532
      - i64-0
    eclass: "31"
    cost: 1000000
  AstExpm1_HerbieType_Math__-2158179654801719225:
    op: AstExpm1_HerbieType_Math__
    children:
      - GetChild__-17074017086827668002
      - TrmOf__-3481525101393754990
    eclass: "56"
    cost: 1
  AstExpm1_HerbieType_Math__-15564543775964812645:
    op: AstExpm1_HerbieType_Math__
    children:
      - GetChild__-1465098945834158911
      - GetChild__-14040062012979708074
    eclass: "60"
    cost: 1
  String-5342629820586423399:
    op: "\"h0\""
    children: []
    eclass: String-5342629820586423399
    cost: 0
  Var-8311271872291325260:
    op: Var
    children:
      - String-5342629820586423399
    eclass: "47"
    cost: 1
  Fma-10050917237229058210:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  Fma-10645116314950871453:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  Fma-7529359344572618602:
    op: Fma
    children: []
    eclass: "5383"
    cost: 1
  Fma-17301992692601812334:
    op: Fma
    children: []
    eclass: "6581"
    cost: 1
  Fma-17135648761937485494:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  Fma-4656402242138846898:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  Fma-9421131195754156095:
    op: Fma
    children: []
    eclass: "1497"
    cost: 1
  Fma-1115809564189179928:
    op: Fma
    children: []
    eclass: "1497"
    cost: 1
  Fma-2770222625774018806:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  Fma-13701072956726874872:
    op: Fma
    children: []
    eclass: "63"
    cost: 1
  NotEq-11964306030268855396:
    op: NotEq
    children: []
    eclass: "84"
    cost: 1
root_eclasses: []
class_data:
  "121":
    type: Proof__
  "119":
    type: Proof__
  "164":
    type: Proof__
  "162":
    type: Proof__
  "218":
    type: Proof__
  "216":
    type: Proof__
  "243":
    type: Proof__
  "241":
    type: Proof__
  "283":
    type: Proof__
  "281":
    type: Proof__
  "332":
    type: Proof__
  "330":
    type: Proof__
  "355":
    type: Proof__
  "353":
    type: Proof__
  "381":
    type: Proof__
  "379":
    type: Proof__
  "509":
    type: Proof__
  "507":
    type: Proof__
  "550":
    type: Proof__
  "548":
    type: Proof__
  "31":
    type: Ast__
  "33":
    type: TrmPrf__
  "25":
    type: Ast__
  "27":
    type: TrmPrf__
  "52":
    type: Ast__
  "54":
    type: TrmPrf__
  "19":
    type: Ast__
  "21":
    type: TrmPrf__
  "13":
    type: Ast__
  "15":
    type: TrmPrf__
  "7":
    type: Ast__
  "9":
    type: TrmPrf__
  "34":
    type: Ast__
  "36":
    type: TrmPrf__
  "1":
    type: Ast__
  "3":
    type: TrmPrf__
  "60":
    type: Ast__
  "62":
    type: TrmPrf__
  "28":
    type: Ast__
  "30":
    type: TrmPrf__
  i64-0:
    type: i64
  "6":
    type: TrmPrf__
  i64-5871781006564002453:
    type: i64
  i64-11743562013128004906:
    type: i64
  i64-17615343019692007359:
    type: i64
  i64-5040379952546458196:
    type: i64
  i64-12574963067145549163:
    type: i64
  Rational-11743562013128004906:
    type: Rational
  "123":
    type: Ast__
  "41":
    type: Ast__
  "95":
    type: Ast__
  "104":
    type: Ast__
  "48":
    type: Ast__
  "56":
    type: Ast__
  "37":
    type: Ast__
  "4":
    type: Ast__
  "10":
    type: Ast__
  "47":
    type: Math
  "40":
    type: HerbieType
  String-5342629820586423399:
    type: String
  "63":
    type: Math
  "4077":
    type: Math
  "2334":
    type: Math
  "2737":
    type: Math
  "5383":
    type: Math
  "1495":
    type: Math
  "1498":
    type: Math
  "11075":
    type: Math
  "6581":
    type: Math
  "416":
    type: Math
  "11369":
    type: Math
  "10512":
    type: Math
  "833":
    type: Math
  "11909":
    type: Math
  "1497":
    type: Math
  "30497":
    type: Math
  "2608":
    type: Math
  "303":
    type: Math
  "2813":
    type: Math
  "39145":
    type: Math
  "9352":
    type: Math
  "84":
    type: Math
  "164617":
    type: Math

